design-proposals/kliup-29-explicit-keys.md

```diff
@@ -0,0 +1,154 @@
+Stream S Created.
+ksal> CREATE STREAM S2 AS SELECT ID, NAME FROM S;
+Key missing from projection. See https://cnfl.io/2LV7ouS.
+The query used to build `S2` must include the key column ROWKEY in its projection.
+```
+
+Understandably, the user may be left wondering where this `ROWKEY` column came from and why they need to
+add it to the projection.
+
+Now that ksqlDB supports more than just `STRING` key columns, it no longer makes sense to add an
+implicit `STRING` key column by default.  Better to let the user define _if_ there is a key column, and if
+so, what it's name and type are.
+
+There is no semantic need for a STREAM to have data in the key, and hence there should be no need for 
+there to be a KEY column defined.  This can also be useful when the data in the key is not in a format 
+ksqlDB can work with, e.g. an Avro, JSON or Protobuf key, or any other format.  It seems unwieldy and 
+confusing to add an implicit `STRING` key column if the actual key is empty or is not something ksqlDB 
+can read.
+
+Unlike STREAMs, TABLEs require a PRIMARY KEY column. Therefore, we should require the user to provide one.  
+
+## What is in scope
+
+* `CREATE TABLE` statements will require a `PRIMARY KEY` column to be defined.
+* `CREATE STREAM` statements without a `KEY` column defined will define a stream without a key column. 
+
+## What is not in scope
+
+* Everything else.
+
+## Value/Return
+
+Removal of implicitly added columns makes statements more declarative. They make it possible to describe 
+a stream with no data, or incompatible data, in the key. They make users _think_ about the name and type
+of the PRIMARY KEY column of their tables.
+
+Combined, these changes make the language easier to use and richer.
+
+## Public APIS
+
+### CREATE TABLE changes
+
+A `CREATE TABLE` statement that does not define a `PRIMARY KEY` column will now result in an error:
+
+```
+ksql> CREATE TABLE INPUT (A INT, B STRING) WITH (...);
+Tables require a PRIMARY KEY. Please define the PRIMARY KEY.
+```
+
+Statements with a `PRIMARY KEY` column will continue to work as before.
+
+Where schema inference is used, the error message will include an example of how to define a partial
+schema to add the primary key:
+
+```ql
+ksql> CREATE TABLE INPUT WITH (value_foramt='Avro', ...);
+Tables require a PRIMARY KEY. Please define the PRIMARY KEY.
+You can define just the primary key and still load the value columns from the Schema registry, for example:
+  CREATE TABLE INPUT (ID INT PRIMARY KEY) WITH (value_foramt='Avro', ...);
+```
+
+### CREATE STREAM changes
+
+A `CREATE STREAM` statement that does not define a `KEY` column will now result in a stream with no
+data being read from the Kafka message's key.
+
+```sql
+CREATE STREAM INPUT (A INT, B STRING) WITH (...);
+-- Results in INPUT with schema: A INT, B STRING, i.e. no key column.
+```
+
+Statements with a `KEY` column will continue to work as before.
+
+## Design
```

Comment:
would be nice to explicitly state in this KLIP what the binary value of the key will be in subsequent queries. e.g.
```sql
CREATE STREAM foo (col VARCHAR) WITH (...); --- there is no declared key, but there is a key in the kafka message
CREATE STREAM bar AS SELECT * FROM foo; --- what is in the binary key of the kafka message?
```
If I understand correctly, the key will still be the same as the key from `foo`, not `null`

Reply:
Good call.

Reply:
From my understanding, the key would be `null`... Because in the first `CS` the key is ignored, it should be "projected away"?

Seems that this aligns with the update to the KLIP @big-andy-coates pushed.